File: 1160/solution.ts
function countCharacters(words: string[], chars: string): number {
  const totalCounts = new Map<string, number>();
  for (const char of chars) {
    totalCounts.set(char, (totalCounts.get(char) ?? 0) + 1);
  }
  let res = 0;
  for (const word of words) {
    const letterCounts = new Map<string, number>();
    for (const char of word) {
      letterCounts.set(char, (letterCounts.get(char) ?? 0) + 1);
    }
    let valid = true;
    for (const key of letterCounts.keys()) {
      if (
        !totalCounts.get(key) ||
        letterCounts.get(key)! > totalCounts.get(key)!
      ) {
        valid = false;
        break;
      }
    }
    res += valid ? word.length : 0;
  }
  return res;
}

File: 1160/solution.java
class Solution {
    public int countCharacters(String[] words, String chars) {
        int[] charCounts = new int[26];
        for (int i = 0; i < chars.length(); i++) {
            charCounts[chars.charAt(i) - 'a']++;
        }
        int res = 0;
        for (String word : words) {
            int[] letterCounts = new int[26];
            for (int i = 0; i < word.length(); i++) {
                letterCounts[word.charAt(i) - 'a']++;
            }
            boolean valid = true;
            for (int i = 0; i < 26; i++) {
                if (charCounts[i] < letterCounts[i]) {
                    valid = false;
                    break;
                }
            }
            res += valid ? word.length() : 0;
        }
        return res;
    }
}